src/vm/dllimport.cpp

```diff
@@ -3719,31 +3719,39 @@ static MarshalInfo::MarshalType DoMarshalReturnValue(MetaSig&           msig,
         else
 #endif // FEATURE_COMINTEROP
         {
-            if (marshalType > MarshalInfo::MARSHAL_TYPE_DOUBLE && IsUnsupportedValueTypeReturn(msig))
-            {
-                if (marshalType == MarshalInfo::MARSHAL_TYPE_BLITTABLEVALUECLASS
-                        || marshalType == MarshalInfo::MARSHAL_TYPE_GUID
-                        || marshalType == MarshalInfo::MARSHAL_TYPE_DECIMAL
+            if (marshalType == MarshalInfo::MARSHAL_TYPE_BLITTABLEVALUECLASS
+                    || marshalType == MarshalInfo::MARSHAL_TYPE_VALUECLASS
+                    || marshalType == MarshalInfo::MARSHAL_TYPE_GUID
+                    || marshalType == MarshalInfo::MARSHAL_TYPE_DECIMAL
 #ifdef FEATURE_COMINTEROP                    
-                        || marshalType == MarshalInfo::MARSHAL_TYPE_DATETIME
+                    || marshalType == MarshalInfo::MARSHAL_TYPE_DATETIME
 #endif // FEATURE_COMINTEROP
-                   )
-                {
-                    if (SF_IsHRESULTSwapping(dwStubFlags))
-                    {
-                        // V1 restriction: we could implement this but it's late in the game to do so.
-                        COMPlusThrow(kMarshalDirectiveException, IDS_EE_NDIRECT_UNSUPPORTED_SIG);
-                    }
-                }
-                else if (marshalType == MarshalInfo::MARSHAL_TYPE_HANDLEREF)
-                {
-                    COMPlusThrow(kMarshalDirectiveException, IDS_EE_BADMARSHAL_HANDLEREFRESTRICTION);
-                }
-                else
+                )
+            {
+                if (SF_IsHRESULTSwapping(dwStubFlags))
                 {
+                    // V1 restriction: we could implement this but it's late in the game to do so.
                     COMPlusThrow(kMarshalDirectiveException, IDS_EE_NDIRECT_UNSUPPORTED_SIG);
                 }
             }
+            else if (marshalType == MarshalInfo::MARSHAL_TYPE_DATE
```

Comment:
These lists of supported/non-supported cases look very arbitrary from use point of view. It is very non-intuitive why DateTime works, but Date does not.

Reply:
This list was pulled from https://github.com/dotnet/coreclr/blob/fdef855fa1df3540ba632bfae850279d627bde66/src/vm/dllimport.cpp#L4013-L4025

Without this, we'd accidentally think we need a return buffer on stdcall x86 and assert when trying to find the entry point for the function.

The part of the code that determines if we need a return buffer for stdcall doesn't have access to the selected MARSHAL_TYPE_* since it hasn't been calculated yet.

Reply:
After doing some more in-depth testing, it actually looks like we can support returning Date types. Since they are `LayoutKind.Auto` and specially marshalled, their EEClass doesn't have a native size calculated, so it's 0. Therefore, `IsUnmanagedValueTypeReturnedByRef` thinks that it is returned by value, which the native representation is. `DateTimeOffset` happens to be in the same situation.